nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp

```diff
@@ -132,23 +133,59 @@ enum RETURN_TYPE {
 /// @brief Base class for Velodyne LiDAR decoder
 class VelodyneScanDecoder
 {
+private:
+  size_t processed_packets_{0};
+  uint32_t prev_packet_first_azm_phased_{0};
+  bool has_scanned_{false};
+
 protected:
+  /// @brief Checks if the current packet completes the ongoing scan.
+  /// TODO: this has been moved from velodyne_hw_interface.cpp and is a temporary solution until
+  /// the Velodyne decoder uses the same structure as Hesai/Robosense
+  /// @param packet The packet buffer to extract azimuths from
+  /// @param packet_seconds The seconds-since-epoch part of the packet's timestamp
+  /// @param phase The sensor's scan phase used for scan cutting
+  void checkAndHandleScanComplete(
+    const std::vector<uint8_t> & packet, int32_t packet_seconds, const uint32_t phase)
+  {
+    if (has_scanned_) {
+      processed_packets_ = 0;
+      reset_pointcloud(packet_seconds);
+    }
+
+    has_scanned_ = false;
+    processed_packets_++;
+
+    // Check if scan is complete
+    uint32_t packet_first_azm = packet[2];  // lower word of azimuth block 0
+    packet_first_azm |= packet[3] << 8;     // higher word of azimuth block 0
+
+    uint32_t packet_last_azm = packet[1102];
+    packet_last_azm |= packet[1103] << 8;
+
+    uint32_t packet_first_azm_phased = (36000 + packet_first_azm - phase) % 36000;
+    uint32_t packet_last_azm_phased = (36000 + packet_last_azm - phase) % 36000;
```

Comment:
Not important - but would be nice to define 1102, 1103, 36000 in velodyne_packet.hpp

Reply:
Addressed in [0064bb8](https://github.com/tier4/nebula/pull/148/commits/0064bb8d6eb129c1b23437eab38db6b25ce2e964)